overlord/devicestate/remodel.go

```diff
@@ -0,0 +1,349 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package devicestate
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/overlord/storecontext"
+)
+
+/*
+
+This is the central logic to setup and mediate the access to the to-be
+device state and dedicated store during remodeling and drive the
+re-registration, leveraging the snapstate.DeviceContext/DeviceCtx and
+storecontext.DeviceBackend mechanisms and also registrationContext.
+
+Different context implementations will be used depending on the kind
+of remodel, and those will play the roles/implement as needed
+snapstate.DeviceContext, storecontext.DeviceBackend and
+registrationContext:
+
+* same brand/model, brand store => updateRemodel
+  this is just a contextual carrier for the new model
+
+* same brand/model different brand store => storeSwitchRemodel this
+  mediates access to device state kept on the remodel change, it also
+  creates a store that uses that and refers to the new brand store
+
+* different brand/model, maybe different brand store => reregRemodel
+  similar to storeSwitchRemodel case after a first phase that performs
+  re-registration where the context plays registrationContext's role
+  (NOT IMPLEMENTED YET)
+
+*/
+
+// RemodelKind designates a kind of remodeling.
+type RemodelKind int
+
+const (
+	// same brand/model, brand store
+	UpdateRemodel RemodelKind = iota
+	// same brand/model, different brand store
+	StoreSwitchRemodel
+	// different brand/model, maybe different brand store
+	ReregRemodel
+)
+
+func (k RemodelKind) String() string {
+	switch k {
+	case UpdateRemodel:
+		return "revision update remodel"
+	case StoreSwitchRemodel:
+		return "store switch remodel"
+	case ReregRemodel:
+		return "re-registration remodel"
+	}
+	panic(fmt.Sprintf("internal error: unknown remodel kind: %d", k))
```

Comment:
I think I'd prefer for String() to always behave and never panic and simply return "unknown... %d" value + plus handle unsupported kinds on different levels (which needs to be done anyway).

Reply:
These values never come from outside.

Reply:
Fwiw we have the same style here:

https://github.com/snapcore/snapd/blob/master/overlord/state/change.go#L104

where arguably values could come from outside but then we would be in  a very confused state anyway.

Reply:
Fair enough, thanks.